File: src/com/brianc/dom/Node.java
package com.brianc.dom;

import java.util.List;
import java.util.Map;

public abstract class Node {
	private List<Node> children;
	final NodeType type;

	public Node(List<Node> children, NodeType type) {
		this.children = children;
		this.type = type;
	}

	public List<Node> getChildren() {
		return children;
	}

	public NodeType getType() {
		return type;
	}

	public static Node createText(String text) {
		return new TextNode(text);
	}

	public static Node createElement(String name, Map<String, String> attributes,
			List<Node> children) {
		return new ElementNode(children, new ElementData(name, attributes));
	}
}

File: src/com/brianc/dom/Node.kt
package com.brianc.dom

abstract class Node(val children: List<Node>, val type: NodeType) {
    companion object {

        fun createText(text: String): Node {
            return TextNode(text)
        }

        fun createElement(name: String, attributes: Map<String, String>,
                          children: List<Node>): Node {
            return ElementNode(children, ElementData(name, attributes))
        }
    }
}
